### ISSUE TITLE
Elapsed build time in console

### ISSUE THREAD

**Issue by mmcc007** [cirruslabs/cirrus-ci-docs#224]
# Description
Include the elapsed time (wall clock difference) of the build

# Context
In general, the most relevant measure to a builder (other than success/failure) is the time it takes to run a build.

I cannot find this value in the console.

...

**Comment by fkorotkov** [comment]
@mmcc007 I'm not quite sure what to do with re-runs here. Elapsed time is not just the duration between build creation time and time when it got into the final state.

There might be a situation when a build fails because of a flaky task. Once the flaky task is re-run and success then the build status changes. How the elapsed build time should look like in this case? What about [recently introduced manual tasks](https://cirrus-ci.org/guide/writing-tasks/#manual-tasks) that can postponed builds completion.

The only viable solution it seems is to do some clever linearization and aggregation of task durations. Imagine you have 2 macOS tasks, 2 Linux tasks and the final manual tasks that deploys everything. All of it for an OSS project with [limits](https://cirrus-ci.org/faq/#are-there-any-limits). Here is now timeline of task execution can look like (each table cell is a time slot):

| Task Name |   |   |   |   |   |   |   |   |
|---|---|---|---|---|---|---|---|---|
| MacOS 1  | * | * |   |   |   |   |   |   |
| MacOS 2  |   |   | * | * |   |   |   |   |
| Linux 1  | * |   |   |   |   |   |   |   |
| Linux 2  | * | * | * |   |   |   |   |   |
| Linux 1 Re-Run  |   |   |   |   |   | * |   |   |
| Deploy  |   |   |   |   |   |   |   | * |
| Overall  | * | * | * | * |   | * |   | * |

In this example Linux tasks are executed in parallel and macOS tasks are sequential due to the concurrency limits. There is also a re-run that happened for *Linux 1* task.

It seems in this case the elapsed time should be 6. The build duration will be still 4 (*Deploy* duration + maximum duration of dependencies).

Does it make sense?

**Comment by mmcc007** [comment]
If you consider build time to be the wall clock time less any time waiting for manual response, you should be able to cover the general case and the manual event case. The manual event case should probably include some extra reporting to user on build time.

If there is a task re-run, you should still have enough info to report the overall build time, but, as in the manual event case, overall build time is not so important in this case.  The duration of the most recent re-run of the task should probably be used in place of the duration of any previous run of the same task.

**Comment by fkorotkov** [comment]
Added clock time calculated by the algorithm described in my previous comment (by merging time intervals of tasks while they were executing).

<img width="517" alt="Screen Shot 2019-05-29 at 9 01 23 PM" src="https://user-images.githubusercontent.com/989066/58601036-1654d480-8255-11e9-8821-c9e2b37f6469.png">

Will see how it goes and if it'll make sense to show the clock duration as the primary duration of a build.

**Comment by mmcc007** [comment]
Looks great!
I assume it includes startup time, etc... (that would be a nice-to-have)

The simplest way to calculate wall clock time is to take a timestamp at start of build and subtract from timestamp at end of build. If a task is re-run, or there is a manual event, their wall clock time can be used to adjust the build wall clock time. (not sure why I didn't mention this already)

Now users can see how long their builds truly take and can more clearly decide when an upgrade to higher limits makes sense.

**Comment by fkorotkov** [comment]
Yeah, clock duration includes scheduling time as well. In the ideal world when you don't hit [the limits](https://cirrus-ci.org/faq/#are-there-any-limits) clock duration should be the same as the regular duration.

**Comment by mmcc007** [comment]
Another cool feature would be a [critical path](https://en.wikipedia.org/wiki/Critical_path_method) report that shows which sequence of tasks took the longest time. Something like:
![image](https://user-images.githubusercontent.com/1980785/58655593-49d44500-82cf-11e9-8de9-f0fcfa04a0db.png)
But that's another issue.

**Comment by fkorotkov** [comment]
@mmcc007 feel free to create an issue! 😉I also think nodes can be rectangles with length proportional to overall execution time. 🤔

**Comment by mmcc007** [comment]
#315
